pkg/federate/federator_test.go

```diff
@@ -0,0 +1,379 @@
+			})
+
+			It("should return an error", func() {
+				Expect(f.Distribute(t.resource)).ToNot(Succeed())
+			})
+		})
+	})
+
+	When("the resource does not exist in the datastore", func() {
+		It("should succeed", func() {
+			Expect(f.Distribute(t.resource)).To(Succeed())
+		})
+	})
+}
+
+func testUpdateStatusFederator() {
+	var (
+		f federate.Federator
+		t *testDriver
+	)
+
+	BeforeEach(func() {
+		t = newTestDiver()
+		t.localClusterID = ""
+	})
+
+	JustBeforeEach(func() {
+		f = federate.NewUpdateStatusFederator(t.dynClient, t.restMapper, t.federatorNamespace)
+		t.resource.SetNamespace(t.targetNamespace)
+		test.CreateResource(t.resourceClient, t.resource)
+	})
+
+	When("the Status is changed", func() {
+		JustBeforeEach(func() {
+			t.resource.Status = corev1.PodStatus{
+				Phase: corev1.PodRunning,
+				PodIP: "10.253.4.6",
+			}
+		})
+
+		It("should update the resource", func() {
+			Expect(f.Distribute(t.resource)).To(Succeed())
+			t.verifyResource()
+		})
+	})
+}
+
+func testDelete() {
+	var (
+		f federate.Federator
+		t *testDriver
+	)
+
+	BeforeEach(func() {
+		t = newTestDiver()
+	})
+
+	JustBeforeEach(func() {
+		f = federate.NewCreateOrUpdateFederator(t.dynClient, t.restMapper, t.federatorNamespace, "")
+	})
+
+	When("the resource exists in the datastore", func() {
+		BeforeEach(func() {
+			existing := t.resource.DeepCopy()
+			existing.SetNamespace(t.targetNamespace)
+			test.CreateResource(t.resourceClient, existing)
+		})
+
+		It("should delete the resource", func() {
+			Expect(f.Delete(t.resource)).To(Succeed())
+
+			_, err := test.GetResourceAndError(t.resourceClient, t.resource)
+			Expect(apierrors.IsNotFound(err)).To(BeTrue())
+		})
+
+		Context("and delete fails", func() {
+			BeforeEach(func() {
+				t.resourceClient.FailOnDelete = apierrors.NewServiceUnavailable("fake")
+			})
+
+			It("should return an error", func() {
+				Expect(f.Delete(t.resource)).ToNot(Succeed())
+			})
+		})
+
+		Context("and no target namespace is specified", func() {
+			BeforeEach(func() {
+				t.federatorNamespace = corev1.NamespaceAll
+				t.resource.SetNamespace(t.targetNamespace)
+			})
+
+			It("should delete the resource from the source namespace", func() {
+				Expect(f.Delete(t.resource)).To(Succeed())
+
+				_, err := test.GetResourceAndError(t.resourceClient, t.resource)
+				Expect(apierrors.IsNotFound(err)).To(BeTrue())
+			})
+		})
+	})
+
+	When("the resource does not exist in the datastore", func() {
+		It("should return NotFound error", func() {
+			Expect(apierrors.IsNotFound(f.Delete(t.resource))).To(BeTrue())
+		})
+	})
+}
+
+type testDriver struct {
+	resource           *corev1.Pod
+	localClusterID     string
+	federatorNamespace string
+	targetNamespace    string
+	keepMetadataFields []string
+	dynClient          *fake.DynamicClient
+	resourceClient     *fake.DynamicResourceClient
+	restMapper         meta.RESTMapper
+	initObjs           []runtime.Object
+}
+
+func newTestDiver() *testDriver {
```

Comment:
```suggestion
func newTestDriver() *testDriver {
```

Super nit :) Maybe you meant newTestDriver()